### ISSUE TITLE
Make all shortcuts customizable

### ISSUE THREAD

**Issue by ElbertCode** [flxzt/rnote#590]
**Is your feature request related to a problem? Please describe.**
If you press Ctrl + ?, a window opens in which several shortcuts are listed, but they are not configurable.

**Describe the solution you'd like**
However, there are a few shortcuts that are customisable, they are listed in the settings. Here you could also list the other shortcuts and make them configurable. To keep the settings from becoming too confusing, you could also create a submenu inside the settings in which you set all the shortcuts. Then you could even make different submenus for the categories of shortcuts (e.g. general (menus, ...), pen, text editing, ...).

**Describe alternatives you've considered**
You could also make the shortcuts configurable right in the ctrl + ? menu, but I would personally prefer the configuration being done in the settings and the menu just showing the current state.

**Additional context**
As for the idea with the submenu, I was inspired by the accessibility settings of the new [GNOME version 44](https://release.gnome.org/44/), which also divide the different categories into submenus.

**Comment by valeratrades** [comment]
People who want configurable shortcuts wouldn't mind a config file, so maybe it could be easier to roll this out to linux only at first, to postpone gui decisions?

Another consideration: if shortcuts are customizable, we probably want to have normal/insert or global/insert scope specification for them, so that it's possible to define say `u` for undo, and not have it clash with text-input mode, while having the main `Ctrl-[0-9]` suite still be accessible there.

If no gui is involved, I could probably make a prototype myself too, if you greenlight it.

**Comment by tornikeo** [comment]
In my case, I just wanted to rebind my xp-pen's keys to something that works in RNote. Since modifying RNote is not yet possible, I instead rebound the xp-pen's buttons, using OpenTabletDriver. Using `xinput` to do this is also possible, but much harder.

For Linux users. Here's how I solved this:
- Install [OpenTabletDriver](https://opentabletdriver.net/Wiki/Install/Linux). 
- Enable driver `systemctl --user enable opentabletdriver.service --now`
- Start the opentabletdriver from start menu

In my case, I have a XP-Pen with 4 auxiliary buttons. I wanted the first button rebound to Ctrl-Z.
- Go to "Auxiliary settings" tab on OpenTabletDriver. 
- Select "Auxiliary Binding 1"
- Three dots to the left -> Advanced binding editor -> Multi-key binding
- Press ctrl-z
- Save, apply, close.

Now it works!

**Comment by MMachado05** [comment]
> People who want configurable shortcuts wouldn't mind a config file, so maybe it could be easier to roll this out to linux only at first, to postpone gui decisions?
> 
> Another consideration: if shortcuts are customizable, we probably want to have normal/insert or global/insert scope specification for them, so that it's possible to define say `u` for undo, and not have it clash with text-input mode, while having the main `Ctrl-[0-9]` suite still be accessible there.
> 
> If no gui is involved, I could probably make a prototype myself too, if you greenlight it.

I don't think this is true at all. Changing default keyboard shortcuts is something anyone might do, and it becomes easier when it's just a matter of clicking some empty box, inputting the key presses you want to have trigger the action, and then waiting for them to register. I think a lot of artists who use things like Krita/Photoshop do this a ton, and they've likely never needed to look at a config file to do so.

Like, I would love to just map letters unused elsewhere (such as 'p') to turn on certain tools, like the pen, rather than have to take my hand off my tablet to use a CTRL-based shortcut.

**Comment by yakeinpoonia** [comment]
when we are expecting this feature to be implemented ?? @flxzt

**Comment by B83C** [comment]
> > People who want configurable shortcuts wouldn't mind a config file, so maybe it could be easier to roll this out to linux only at first, to postpone gui decisions?
> > Another consideration: if shortcuts are customizable, we probably want to have normal/insert or global/insert scope specification for them, so that it's possible to define say `u` for undo, and not have it clash with text-input mode, while having the main `Ctrl-[0-9]` suite still be accessible there.
> > If no gui is involved, I could probably make a prototype myself too, if you greenlight it.
> 
> I don't think this is true at all. Changing default keyboard shortcuts is something anyone might do, and it becomes easier when it's just a matter of clicking some empty box, inputting the key presses you want to have trigger the action, and then waiting for them to register. I think a lot of artists who use things like Krita/Photoshop do this a ton, and they've likely never needed to look at a config file to do so.
> 
> Like, I would love to just map letters unused elsewhere (such as 'p') to turn on certain tools, like the pen, rather than have to take my hand off my tablet to use a CTRL-based shortcut.

Yes, was expecting letters based shortcut keys, but turned out to be missing :[

**Comment by robearista-cpu** [comment]
I really like rnote, but want to be able to change brush colors with shortcuts. I just want something like press a button on your keyboard to make your brush turn red. In the far future, maybe include shortcuts for cycling colors forward and backward. I'm a big fan of color switching with a button, it helps me color coordinate my notes.

**Comment by der-wasguckstdu-dreier** [comment]
I also have the need of custom keyboard hotkeys. Im on fedora linux and as obsidian is running through electron with the old wayland (sorry I am not a professional :D) I needed to remap a stylus button in the KDE settings to E. The reason for this workaround is that  Electron does not forward both of my wacom pen buttons. Having my pen button remapped to E, it isnt recognized by rNote properly. I dont want to remap the pen each time before taking notes for obvious reasons...

**Comment by jncrenshaw** [comment]
CTRL+1-5 is crazytown for tool selection.
If you're going to make default hotkeys that bad, at least make them configurable.
I am very confused by why this hasn't been implemented, even years later...

**Comment by Kneemund** [comment]
> CTRL+1-5 is crazytown for tool selection.
> If you're going to make default hotkeys that bad, at least make them configurable.
> I am very confused by why this hasn't been implemented, even years later...

@jncrenshaw It seems like you didn't notice yet: this is free and open source software. You are free to make suggestions regarding the current shortcuts, and you are free to spend your own time on implementing and contributing this feature in a (non-slop) pull request. It's likely that someone would even donate their spare time to review your changes :)
